### PR TITLE
Quota enable config to user/domain scope 

### PR DESCRIPTION
**Pull request by BryanMLima**
### Description

Currently, ACS only has the option to enable the Quota plugin globally. However, for some use cases having the plugin active to specific accounts and domains may be more interesting. Therefore, this PR aims to add this functionality through the quota config `quota.account.enabled`. Moreover, a new column was added to the quota summary section in the UI showing if the account has the plugin enabled. In the future, a filter will be added for accounts with the plugin enabled, disabled and for removed accounts.

Finally, for the config to be used in a domain scope, when it cannot be found at user level, it is required that the following global configurations are enabled: `enable.account.settings.for.domain` and `enable.domain.settings.for.child.domain`.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

### Screenshots (if appropriate):

List of accounts with the plugin enabled and disabled:
image

### How Has This Been Tested?
This was tested in a local lab setting the `quota.account.enabled` to true and false to random accounts. In the quota summary section the accounts where correctly displayed: accounts with the plugin disabled had the `(quota-disabled)` string attached and accounts with the plugin enabled were normally displayed.

### CODE REVIEW

File: framework/quota/src/main/java/org/apache/cloudstack/quota/dao/QuotaAccountDaoImpl.java [hunk 1]
```diff
@@ -36,7 +38,15 @@ public class QuotaAccountDaoImpl extends GenericDaoBase<QuotaAccountVO, Long> im
 
     @Override
     public List<QuotaAccountVO> listAllQuotaAccount() {
-        return listAllQuotaAccount(null, null).first();
+        List<QuotaAccountVO> accountsWithQuotaEnabled = new ArrayList<>();
+        for (QuotaAccountVO account : listAllQuotaAccount(null, null).first()) {
+            if (getQuotaAccountEnabled(account.getAccountId())) {
```
Comment (DaanHoogland) [inline framework/quota/src/main/java/org/apache/cloudstack/quota/dao/QuotaAccountDaoImpl.java:43]:
```suggestion
            if (Boolean.TRUE.equals(getQuotaAccountEnabled(account.getAccountId()))) {
```

**Review by JoaoJandre — APPROVED** [review]
CLGTM

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]

File: framework/quota/src/main/java/org/apache/cloudstack/quota/QuotaManagerImpl.java [hunk 2]
```diff
@@ -133,9 +134,21 @@ public boolean stop() {
...
+    public boolean isQuotaEnabledForAccountDomain(final AccountVO account) {
+        boolean isQuotaAccountEnabled = QuotaConfig.QuotaAccountEnabled.valueIn(account.getAccountId());
+        if (!isQuotaAccountEnabled) {
+            s_logger.debug(String.format("Considering usage records as calculated and skipping it because the account [%s] has the quota plugin disabled.", account));
+        }
+        return isQuotaAccountEnabled;
+    }
+
+    public boolean isUsageRecordsEmpty(final Pair<List<? extends UsageVO>, Integer> usageRecords) {
+        return usageRecords == null || usageRecords.first() == null || usageRecords.first().isEmpty();
```
Comment (GutoVeronezi) [inline framework/quota/src/main/java/org/apache/cloudstack/quota/QuotaManagerImpl.java:146]: We can use `org.apache.commons.collections.CollectionUtils#isEmpty` here.
```suggestion
        return usageRecords == null || CollectionUtils.isEmpty(usageRecords.first());
```

File: framework/quota/src/main/java/org/apache/cloudstack/quota/dao/QuotaAccountDaoImpl.java [hunk 3]
```diff
@@ -36,7 +38,15 @@ public class QuotaAccountDaoImpl extends GenericDaoBase<QuotaAccountVO, Long> im
 
     @Override
     public List<QuotaAccountVO> listAllQuotaAccount() {
-        return listAllQuotaAccount(null, null).first();
+        List<QuotaAccountVO> accountsWithQuotaEnabled = new ArrayList<>();
+        for (QuotaAccountVO account : listAllQuotaAccount(null, null).first()) {
+            if (Boolean.TRUE.equals(getQuotaAccountEnabled(account.getAccountId()))) {
```
Comment (GutoVeronezi) [inline framework/quota/src/main/java/org/apache/cloudstack/quota/dao/QuotaAccountDaoImpl.java:43]: We can use `org.apache.commons.lang3.BooleanUtils#isTrue()` here: 

```suggestion
            if (BooleanUtils.isTrue(getQuotaAccountEnabled(account.getAccountId()))) {
```

File: ui/src/config/section/plugin/quota.js [hunk 4]
```diff
@@ -28,7 +28,10 @@ export default {
       title: 'label.quota.summary',
       icon: 'bars-outlined',
       permission: ['quotaSummary'],
-      columns: ['account', 'domain', 'state', 'currency', 'balance', 'quota'],
+      columns: [
+        {
+          account: (record) => record.quotaenabled ? record.account : record.account + ' (quota-disabled)'
```
Comment (GutoVeronezi) [inline ui/src/config/section/plugin/quota.js:33]: @BryanMLima, would not be better to add it as a column?

File: framework/quota/src/main/java/org/apache/cloudstack/quota/dao/QuotaAccountDaoImpl.java (hunk 3 above)
Comment (DaanHoogland) [reply framework/quota/src/main/java/org/apache/cloudstack/quota/dao/QuotaAccountDaoImpl.java:43]: @GutoVeronezi how is this different (/better) ?

Comment (GutoVeronezi) [reply framework/quota/src/main/java/org/apache/cloudstack/quota/dao/QuotaAccountDaoImpl.java:43]: @DaanHoogland nothing in special in this case, just demonstrating a library that facades the operation (better to read IMHO).

**Comment by DaanHoogland** [comment]
@BryanMLima can you look at the conflicts

File: framework/quota/src/main/java/org/apache/cloudstack/quota/QuotaManagerImpl.java [hunk 5]
```diff
@@ -327,6 +327,13 @@ protected List<QuotaUsageVO> createQuotaUsagesAccordingToQuotaTariffs(AccountVO
                     continue;
                 }
 
+                if (!QuotaConfig.QuotaAccountEnabled.valueIn(account.getAccountId())) {
```
Comment (DaanHoogland) [inline framework/quota/src/main/java/org/apache/cloudstack/quota/QuotaManagerImpl.java:330]:
```suggestion
                if (!Boolean.TRUE.equals(QuotaConfig.QuotaAccountEnabled.valueIn(account.getAccountId()))) {
```

Comment (DaanHoogland) [reply framework/quota/src/main/java/org/apache/cloudstack/quota/QuotaManagerImpl.java:330]: also join this conditional with the preceding one. The code blocks are the same.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by stephankruggg — APPROVED** [review]

**Comment by DaanHoogland** [comment]
@GutoVeronezi , I trust all your comments have been met?

**Review by GutoVeronezi — APPROVED** [review]

**Comment by GutoVeronezi** [comment]
@DaanHoogland, yes, they have been met.

I will merge this one based on the packaging and tests results  (https://github.com/apache/cloudstack/pull/6690#issuecomment-1291608946 and https://github.com/apache/cloudstack/pull/6690#issuecomment-1292631838) and on the approvals.

**Pull request merged**